### PR TITLE
feat: collateral per slot

### LINKED ISSUE #386: [marketplace] support slot collateral
The changes in https://github.com/status-im/codex-contracts-eth/pull/44 introduce slot collateral. The client needs to be updated to support these changes, as well as the sample code in `contracts/readme.md`.

### PR DESCRIPTION
**Pull request by AuHau**
Implements the collateral per slot functionality.
Also tweaks the testing setup with removing the token minting functionality (pretty much removed the `TestToken`) as tokens are now minted directly in Hardhat. This lead to the surfacing of a few bugs that I thank to @markspanbroek that helped me to overcome them. 

Closes #386

### CODE REVIEW

File: codex/purchasing/states/pending.nim [hunk 1]
```diff
@@ -10,7 +10,6 @@ method enterAsync(state: PurchasePending) {.async.} =
     raiseAssert "invalid state"
 
   try:
-    await purchase.market.approveFunds(request.price())
```
Comment (markspanbroek) [inline codex/purchasing/states/pending.nim:13]: Why is this removed?

**Review by markspanbroek — COMMENTED** [review]
Beautiful! I really like how much boilerplate is removed from the tests, now that you added minting of coins to the hardhat deployment.

File: openapi.yaml [hunk 2]
```diff
@@ -137,6 +138,9 @@ components:
           type: number
           description: Additional number of nodes on top of the `nodes` property that can be lost before pronouncing the content lost
           default: 0
+        collateral:
+          type: string
+          description: Hexadecimal encoded number that represents how much
```
Comment (markspanbroek) [inline openapi.yaml:143]: Missing end of sentence

File: tests/integration/twonodes.nim [hunk 3]
```diff
@@ -49,6 +49,8 @@ template twonodessuite*(name: string, debug1, debug2: bool, body) =
         "--eth-account=" & $accounts[1]
       ], debug = debug2)
 
+
```
Comment (markspanbroek) [inline tests/integration/twonodes.nim:52]: Only empty lines?

File: tests/integration/testIntegration.nim [hunk 4]
```diff
@@ -4,10 +4,12 @@ import pkg/stint
 import ../contracts/time
 import ../codex/helpers/eventually
 import ./twonodes
-import ./tokens
 
 twonodessuite "Integration tests", debug1 = false, debug2 = false:
 
+  setup:
+    await provider.advanceTime(1.u256)
```
Comment (markspanbroek) [inline tests/integration/testIntegration.nim:11]: This probably requires a comment explaining why it's here :smile:

**Review by markspanbroek — DISMISSED** [review]

**Review by emizzle — APPROVED** [review]
Looks great Adam, well done! I've created a new issue for updating `contracts/readme.md` as it wouldn't be appropriate to ask that to be done in this PR: https://github.com/status-im/nim-codex/issues/394.
